pm: domains: Alternate power domain implementation

JordanYates opened the pull request:
This PR is an alternate implementation (of details) of power domains from the work presented in #39344.

The primary differences are:
1. Power domain drivers are responsible for notifying child devices of state changes themselves.
2. Custom drivers can be notified and handle transition failures if they wish.
3. Power domain transitions are integrated into the existing PM actions, instead of a new set of actions that can occur asynchronously.

This pulls the power domain logic out of the common `pm_device_action_run` function.

Commit:
pm: Add power domain infra structure

Add support for power domains on Zephyr. Power domains are implemented
as simple devices so they can use the existent Zephyr API, for resume
and suspend sync and async and also reference count.

The pm subsystem will ensure that domains are resumed before and
suspended after devices using them. For device runtime power
management, every time the device is got or released the same actions
is done to the domain it belongs.

As domains are implemented as simple devices, it is totally acceptable
a domain belongs to another domain.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
cc13xx-cc26x2: Add power domains

Add power domains infra-structure. Power domains are automatically
generated when declared in DT and bound to devices that depends on
them.

Power domains just need a resource id associated to them that is used
in to power on/off them using TI HAL.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Add power domain information

Document power domains on Zephyr.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device_runtime: Use pm_device_action_run

Use pm_device_action_run instead of calling the pm action callback
directly because pm_device_action_run accounts power domains state
transitions.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Make power domain optional

Add a Kconfig symbol to enable/disable power domain on Zephyr.
Disabling power domain save some memory / space.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: add action to turn device on

Adds `PM_DEVICE_ACTION_TURN_ON` to transition from `PM_DEVICE_STATE_OFF`
to `PM_DEVICE_STATE_SUSPENDED`.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
pm: device: helper for running actions on children

Adds a helper function for running PM actions on child devices. A custom
implementation of `device_supported_foreach` is used as we don't
necessarily want the early exit behaviour of that function in the power
domain context.

The early exit behaviour can be obained through a return value in the
failure callback if that is desired.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
doc: pm: power_domain: update behaviour

Update the power domain docs with the alternate behaviour in this PR.
This is a separate commit to illustrate the differences and should be
squashed into the original commit before merging.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
cc13x2-cc26x2: power: update impl

Update the power domain implementation to illustrate the differences
in this PR's approach. This commit should be squashed before merging.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
@gmarull

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -83,15 +90,19 @@ static void runtime_suspend_work(struct k_work *work)
-	if (ret == 0) {
-		pm->state = PM_DEVICE_STATE_SUSPENDED;
-	}
+	ret = pm_device_action_run(pm->dev, PM_DEVICE_ACTION_SUSPEND);
```
note: I assume this could be left as it was now?

JordanYates replied:
I believe so, but I'm not entirely clear why it was changed in the first place

gmarull commented on doc/guides/pm/power_domain.rst:
```diff
@@ -0,0 +1,184 @@
+
+.. code-block:: c
+
+    static int mydomain_pm_action(const struct device *dev,
+                               enum pm_device_action *action)
+    {
+        switch (action) {
+        case PM_DEVICE_ACTION_RESUME:
+            /* resume the domain */
+            ...
+            /* notify children domain is now powered */
+            pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_ON, NULL);
+            break;
+        case PM_DEVICE_ACTION_SUSPEND:
+            /* notify children domain is going down */
+            pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
+            /* suspend the domain */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_ON:
+            /* turn on the domain (e.g. setup control pins to disabled) */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_OFF:
+            /* turn off the domain (e.g. reset control pins to default state) */
+            ...
+            break;
+        default:
+            return -ENOTSUP;
+        }
+
+        return 0;
+    }
```
could the domain workflow be clarified now? When/how is the domain turned off/on, and what does this imply? what are the differences between resume/suspend/turn_on/turn_off?

ceolin replied:
AFAIU these operations are only triggered when the domain is nested in another domain that was suspended / resumed.

gmarull commented on subsys/pm/device.c:
```diff
@@ -133,6 +140,36 @@ int pm_device_action_run(const struct device *dev,
+		device_handle_t dh = handles[i];
+		const struct device *cdev = device_from_handle(dh);
+
+		rc = pm_device_action_run(cdev, action);
+		if (failure_cb && rc < 0) {
```
maybe `-ENOTSUP` should require special handling (if action if optional should not be treated as an error)

JordanYates replied:
I would prefer to leave special handling to the callback, maybe someone wants a strict regulator that warns if you put a sensor on it that doesn't handle power up/down

gmarull commented on subsys/pm/device.c:
```diff
@@ -119,6 +119,13 @@ int pm_device_action_run(const struct device *dev,
+			return -EALREADY;
+		}
+
+		state = PM_DEVICE_STATE_SUSPENDED;
```
note: according to docs, not implementing support for TURN_ON should not have any implications, but it has as it is now: action_cb will return -ENOTSUP and so suspended state will not be applied (see L134)

gmarull replied:
tldr; we need some unit tests that check API behavior

ceolin replied:
same thing for `TURN_OFF` now that is used by a power domain to notify children.

gmarull reviewed (commented):
I like the ideas brought in this proposal, it will help on making progress. I have a few questions.

ceolin commented:
> I like the ideas brought in this proposal, it will help on making progress. I have a few questions.

The turn on / off is a good idea.

Notify children is not a basic behavior for all domains?? We simplify the subsystem logic but we put more responsibility into the domain implementation. Not bad though. 

I am looking it from my cel and will only be able to properly review tomorrow night

JordanYates commented:
> Notify children is not a basic behavior for all domains?? We simplify the subsystem logic but we put more responsibility into the domain implementation. Not bad though.

Notifying children is a basic behaviour, but what to do when one fails is not. I can easily see custom regulator drivers being written that will accept any device, some devices, or no devices failing. This flexibility is lacking when the return values are ignored because they are buried in `device_pm_get`.

I don't count putting more responsibility on the domain implementation as a bad thing. Its a power domain driver, asking it to call a function to notify children does not seem like too large a burden.

Commit:
pm: device: helper to query power domain

Adds a helper for devices to check if they are on a power-domain or not.
Drivers can use this information to determine if they will be turned on
at some point in the future, or if they are already turned on.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
pm: device_runtime: init into PM_DEVICE_STATE_OFF

Add a function to tell runtime power management that the device is
starting in the off state instead of active or suspended.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
drivers: power_domain: gpio controlled domain

Initial implementation of a simple GPIO controlled power domain.
It exposes no API of its own, all functionality is contained inside
the runtime power management callbacks.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: pm: power_domains: initial demo

Add an initial demonstration of a physical power domain running on the
`qemu_cortex_m3` platform.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
I've added the simplest version of a `power-domain` device and a `qemu` test which cycles some dependent devices.
```
/ {
	test_reg_0: test_reg_0 {
		compatible = "power-domain-gpio";
		label = "test-reg-0";
		enable-gpios = <&gpio0 0 0>;
	};

	test_reg_1: test_reg_1 {
		compatible = "power-domain-gpio";
		label = "test-reg-1";
		enable-gpios = <&gpio0 1 0>;
	};

	test_reg_chained: test_reg_chained {
		compatible = "power-domain-gpio";
		label = "test-reg-chained";
		enable-gpios = <&gpio0 2 0>;
		power-domain = <&test_reg_0>;
	};
};
```
Which shows the sequencing operating correctly.
```
START - test_demo
Enabling runtime power management on regulators
Cycling: test-reg-0
I: test-reg-0 is now ON
I: test-reg-chained is OFF and powered
I: test-reg-chained is OFF and not powered
I: test-reg-0 is now OFF and powered
Cycling: test-reg-1
I: test-reg-1 is now ON
I: test-reg-1 is now OFF and powered
Cycling: test-reg-chained
I: test-reg-0 is now ON
I: test-reg-chained is OFF and powered
I: test-reg-chained is now ON
I: test-reg-chained is now OFF and powered
I: test-reg-chained is OFF and not powered
I: test-reg-0 is now OFF and powered
DONE
```
One thing this exercise has exposed is that correctly implementing devices on power domain depends on making an assumption about the state of the domain by default. If the domain is on by default, you need to run `PM_DEVICE_ACTION_TURN_ON` code in your init function, because you are already powered up. If it is off by default you can't , because the device has no power.

I have made the assumption here that all power domains are turned off until explicitly turned on.
Given how the implementation works, it may even make sense for the drivers to depend on `PM_DEVICE_RUNTIME`.

ceolin commented:
> > Notify children is not a basic behavior for all domains?? We simplify the subsystem logic but we put more responsibility into the domain implementation. Not bad though.
> 
> Notifying children is a basic behaviour, but what to do when one fails is not. I can easily see custom regulator drivers being written that will accept any device, some devices, or no devices failing. The most likely custom implementation is one that will accept any device failing that isn't the device that was asked for. This flexibility is lacking when the return values are ignored because they are buried in `device_pm_get`.
> 
> I don't count putting more responsibility on the domain implementation as a bad thing. Its a power domain driver, asking it to call a function to notify children does not seem like too large a burden.



> > Notify children is not a basic behavior for all domains?? We simplify the subsystem logic but we put more responsibility into the domain implementation. Not bad though.
> 
> Notifying children is a basic behaviour, but what to do when one fails is not. I can easily see custom regulator drivers being written that will accept any device, some devices, or no devices failing. The most likely custom implementation is one that will accept any device failing that isn't the device that was asked for. This flexibility is lacking when the return values are ignored because they are buried in `device_pm_get`.

I got it. The responsibility to notify children was transferred to domains implementation because of the error handling. I thought about it and to to integrate it into the subsystem would require change the device API that I was avoiding.

I really would like to have this behavior done automatically for power domains, but between change an API that is used by all devices or make it responsibility of the power domain driver I prefer the latter.
 
> 
> I don't count putting more responsibility on the domain implementation as a bad thing. Its a power domain driver, asking it to call a function to notify children does not seem like too large a burden.

It is not indeed, but it is not common pattern in the project. Anyway, I don't think it is bad, I rather have it and optimize later than don't have it.

ceolin commented on doc/guides/pm/power_domain.rst:
```diff
@@ -4,21 +4,20 @@ Power Domain
+powered by a common source to be notified of power source state changes in a
+generic fashion. Application code that is using device A does not need to know
+that device B is on the same power domain and should also be configured into a
+low power state.
```
I am finding this very last sentence confuse (or not necessary). The "and should also be configured into a low power state".

ceolin commented on doc/guides/pm/power_domain.rst:
```diff
@@ -4,21 +4,20 @@ Power Domain
-``compatible`` node's property.
+option :kconfig:`PM_DEVICE_POWER_DOMAIN` has to be set.
+
+When a power domain turns itself on or off, it is the responsibilty of the
```
the `on` or `off` words here may sound confuse since the domain notify children when it is suspended / resumed using `on` and `off` actions. Maybe replace them with `active` or `suspended` ?

ceolin commented:
@JordanYates  @gmarull  How you want to proceed with it ? I'm ok with this solution, there are small clarifications and adjusts to be done, but nothing blocker for me.

@JordanYates do you want to move forward and move from the DRAFT stage ?

@gmarull Anything that would make you block it ?

Power domains are being discussed for, at least three months, nothing really major appeared, so I'm really pushing to have it before the next release. This solution move a lot of responsibility from subsystem to drivers, so the impact should not be big. Also, the feature is optional and should not have big impact in people not using it.

JordanYates commented:
> @JordanYates do you want to move forward and move from the DRAFT stage ?

I am happy for you to roll in/squash any of my commits to your PR if you want to continue there. If pushing this feature becomes my responsibility it will be pretty low on my priority list, given that I will probably be on the LTS for quite some time.

ceolin commented:
> > @JordanYates do you want to move forward and move from the DRAFT stage ?
> 
> I am happy for you to roll in/squash any of my commits to your PR if you want to continue there. If pushing this feature becomes my responsibility it will be pretty low on my priority list, given that I will probably be on the LTS for quite some time.

Ok, I'll move forward and squash these changes there.

ceolin commented:
Closing it since changes in this pr were included in the original pr. Thanks for the help / work.
